### PR TITLE
Add comprehensive JSHint config

### PR DESCRIPTION
**Pull request by myabc**

### CODE REVIEW

**Comment by TeatroIO** [comment]
I've prepared a stage. <a href='http://feature-jshint-config-with-fixes.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Click to open.</a>

**Comment by 0xF013** [comment]
Hound complained about defining inner function not at the top of the outer function. They are hoisted anyway so it allows for stuff like this:

```
.directive('name', function() {
  return {
    templateUrl: "...",
    link: link,
    controller: controller
  };

  function link() { ... }
  function controller() { ... }
});
```

This way you can see the definition of the component right away instead of scrolling to the bottom for the directive's return statement. Is it possible to make jshint ignore this case?

**Comment by myabc** [comment]
@0xF013 we could change the config for `latedef`. I agree your directive example does make the intent clearer, ~~although we're not actually writing code like this anywhere~~. I'm not sure this outweighs the other disadvantages of hoisting though.

**Comment by myabc** [comment]
@0xF013 Correction: I see you're using hoisting in the `inplaceEditor` directive.

**Comment by 0xF013** [comment]
@myabc I have moved the functions definition to top on the editable description branch

**Comment by myabc** [comment]
For reference: https://github.com/johnpapa/angularjs-styleguide#ng-annotate (thanks @0xF013!)

**Comment by myabc** [comment]
@opf/developers @0xF013 won't have time to write up a JS style guide, so please review as is. Opinions welcome.

File: .jshintrc [hunk 1]
```diff
@@ -1,3 +1,54 @@
 {
-  "sub": true
+  "node": true,
```
Comment (myabc) [inline .jshintrc:2]: Refer to http://www.jshint.com/docs/options/

Comment (hschink) [inline .jshintrc:2]: Which means that we actually need to configs (one for server side and one for the client side JS code). :wink:

File: .jshintrc [hunk 2]
```diff
@@ -1,3 +1,54 @@
...
+  "immed": true,
+  "latedef": "nofunc",
+  "newcap": true,
+  "noarg": true,
+  "noempty": true,
+  "nonew": true,
+  "maxlen": 100,
+  "quotmark": true,
+  "sub": true,
+  "trailing": true,
```
Comment (hschink) [inline .jshintrc]: I haven't found this one in the [JSHint documentation](http://www.jshint.com/docs/options/).

**Comment by hschink** [comment]
@myabc I'll merge the branch as soon as https://github.com/opf/openproject/pull/2184#discussion_r20347862 is answered and the branch is rebased.

File: .jshintrc (hunk 2 above)
Comment (myabc) [inline .jshintrc]: @hschink good find. It looks like the option removed. Fixed in c6e58bf.

File: .jshintrc (hunk 1 above)
Comment (myabc) [inline .jshintrc:2]: @hschink AFAICT two JSHint configs is not something that is supported by Hound :disappointed: 

But… `node` is actually correct in this case. As of 66f796c0, we now use Common JS (node-style) modules and `require` statements in our front-end code. Webpack parses these before building JS bundles.

**Pull request merged**